Move ironic into cluster

stbenjam opened the pull request:
This deploys the Ironic containers in the cluster itself, rather than
the provisioning host. The provisioning host still maintains an httpd
server for hosting a cache of the IPA images, as well as any additional
images the user wants to provision with (by default we populate it with
CentOS).

Goes with https://github.com/metal3-io/baremetal-operator/pull/309

fixes #41

hardys commented:
> Thoughts?

Unless the downloading needs to be cluster-hosted as well it seems potentially simpler to download directly onto the host, then either run httpd natively, or run a container with httpd that mounts the host directory with the images?

stbenjam commented on 03_launch_mgmt_cluster.sh:
```diff
@@ -27,8 +27,8 @@ M3PATH="${GOPATH}/src/github.com/metal3-io"
-BMOREPO="${BMOREPO:-https://github.com/metal3-io/baremetal-operator.git}"
-BMOBRANCH="${BMOBRANCH:-master}"
+BMOREPO="${BMOREPO:-https://github.com/stbenjam/baremetal-operator.git}"
+BMOBRANCH="${BMOBRANCH:-ironic-deployment}"
```
Temporarily to run the changes from BMO

stbenjam commented:
> > Thoughts?
> 
> Unless the downloading needs to be cluster-hosted as well it seems potentially simpler to download directly onto the host, then either run httpd natively, or run a container with httpd that mounts the host directory with the images?

We want the IPA images to be downloaded in the cluster though, right? I made the ipa-image-downloader an init container. I feel like it's kind of nice that all those bits are all hosted in one place.

Could we run two different http servers for metal3-dev-env? The one on the provisioning host would be for a user to dump any provisioning images they want in (and we prepopulate it with CentOS).

hardys commented:
> @hardys We want the IPA images to be downloaded in the cluster though, right?

Well, it's not essential but yes - note we can use CACHEURL to point the ipa downloader container at the host, so we don't download the RDO tarball every run which will save some time/bandwidth.

stbenjam commented:
Ah good point about CACHEURL. I've put the httpd container back on the provisioning host, so we can grab centos there. IPA is still downloaded from within the cluster.  I think this looks closer to how we did things with OpenShift.

Should be ready for a look, along with https://github.com/metal3-io/baremetal-operator/pull/309

russellb commented on 01_prepare_host.sh:
```diff
@@ -84,13 +99,17 @@ function init_minikube() {
           --type network --config
+      sudo virsh attach-interface --domain minikube \
+          --model virtio --source baremetal \
+          --type network --config
```
Do you know why this was not needed before?  I don't have this, yet I'm using the baremetal network to ssh to my provisioned machines.

```
$ sudo virsh net-dhcp-leases baremetal
 Expiry Time          MAC address        Protocol  IP address                Hostname        Client ID or DUID
-------------------------------------------------------------------------------------------------------------------
 2019-09-30 15:57:38  00:8d:ab:f6:f2:5b  ipv4      192.168.111.21/24         worker-0        -

 $ fssh centos@192.168.111.21
Warning: Permanently added '192.168.111.21' (ECDSA) to the list of known hosts.
Last login: Mon Sep 30 19:16:41 2019 from gateway
[centos@worker-0 ~]$ 
```

stbenjam replied:
I'll investigate to see exactly why it was needed, but Ironic when running in the cluster couldn't reach vbmc without it

stbenjam replied:
I'll investigate to see exactly why it was needed, but Ironic when running in the cluster couldn't reach vbmc without it

stbenjam replied:
I definitely need the interface for Ironic in the cluster to talk to vbmc.  Otherwise Ironic says:

```
| last_error             | Failed to get power state for node c069fe7d-6d38-407e-9206-8745f19df274. Error: IPMI call failed: power status.                                                                                                                                                                       ```

When I do this from the Ironic container:

```
[root@minikube /]# ipmitool -I lanplus -H 192.168.111.1 -p 6230 -Uadmin -Ppassword power status
```

The request comes over the interface, I see it with tcpdump, but vbmc never replies...

```
09:23:02.541681 IP (tos 0x0, ttl 64, id 41227, offset 0, flags [DF], proto UDP (17), length 51)
    192.168.122.224.50637 > 192.168.111.1.6230: UDP, length 23
09:23:02.541701 IP (tos 0x0, ttl 64, id 41227, offset 0, flags [DF], proto UDP (17), length 51)
    192.168.122.224.50637 > 192.168.111.1.6230: UDP, length 23
```

russellb commented:
/test-integration-centos

russellb commented:
/test-centos-integration

russellb commented on 01_prepare_host.sh:
```diff
@@ -60,15 +76,14 @@ if [ ! -f ${CENTOS_IMAGE} ] ; then
 fi
 popd
 
+sudo "${CONTAINER_RUNTIME}" wait -i 1000 ipa-downloader
```
looks like this doesn't work when CONTAINER_RUNTIME is `docker`

https://jenkins.nordix.org/job/airship_metal3io_metal3_dev_env_integration_test_centos/13/console

stbenjam replied:
I did check to make sure docker had wait, but it doesn't support `-i`, I don't think we really need to specify the interval.

stbenjam commented:
/test-centos-integration

elfosardo commented on 01_prepare_host.sh:
```diff
@@ -38,20 +38,36 @@ if ! which kustomize 2>/dev/null ; then
 
-# Download Ironic binary and Ironic inspector image
+# Clean-up any old ironic containers
+for name in ironic ironic-inspector dnsmasq httpd mariadb ipa-downloader; do
```
wondering if we should just add a variable with all those services in common.sh since we reuse them in host_cleanup.sh

stbenjam replied:
Most of the code you're commenting here is existing, I've just moved it.  I think they're good suggestions but I'm not keen to rewrite everything here as part of this change.

stbenjam replied:
Actually I'll fix what I can here

elfosardo replied:
of course can be done in follow-ups

elfosardo commented on 01_prepare_host.sh:
```diff
@@ -38,20 +38,36 @@ if ! which kustomize 2>/dev/null ; then
-# Download Ironic binary and Ironic inspector image
+# Clean-up any old ironic containers
+for name in ironic ironic-inspector dnsmasq httpd mariadb ipa-downloader; do
+    sudo "${CONTAINER_RUNTIME}" ps | grep -w "$name$" && sudo "${CONTAINER_RUNTIME}" kill $name
```
maybe use pgrep instead of ps | grep ?
Also I advise against using name for killing, I would use pid

stbenjam replied:
This is docker ps (or podman ps) not ps itself

elfosardo commented on 01_prepare_host.sh:
```diff
@@ -38,20 +38,36 @@ if ! which kustomize 2>/dev/null ; then
+  if  sudo "${CONTAINER_RUNTIME}" pod exists ironic-pod ; then
+      sudo "${CONTAINER_RUNTIME}" pod rm ironic-pod -f
+  fi
+  # Create pod
```
let's pause here? At least a 1 second sleep would be nice, just to give the time to the system to process the precedent change

stbenjam replied:
This is `docker ps` (or `podman ps`) not ps itself

stbenjam replied:
Why sleep? dev-scripts doesn't sleep for the same code, and I don't see any evidence `podman pod rm` is asynchronous

hardys replied:
Yeah I agree, random sleeps in scripts are poor practice IMHO, if there's something we actually need to wait for then we should do a loop, podman wait or whatever, but as mentioned we don't wait in dev-scripts so this is probably fine as-is.

`man podman-pod-rm` doesn't actually document the exit status, but podman-rm says exit code zero if all containers are removed, so I'd assume podman pod rm works the same, e.g we should fail if the pod isn't removed, assuming set -e

elfosardo replied:
I'm totally up for a more robust approach
again, can be done in a follow up

elfosardo commented on 01_prepare_host.sh:
```diff
@@ -38,20 +38,36 @@ if ! which kustomize 2>/dev/null ; then
-pushd "$IRONIC_DATA_DIR/html/images"
-if [ ! -f ironic-python-agent.initramfs ]; then
-    curl --insecure --compressed -L https://images.rdoproject.org/master/rdo_trunk/current-tripleo-rdo/ironic-python-agent.tar | tar -xf -
+if [[ "${CONTAINER_RUNTIME}" == "podman" ]]; then
```
if we're planning to use different runtimes I suggest to convert this to case instead of if

elfosardo commented on 01_prepare_host.sh:
```diff
@@ -38,20 +38,36 @@ if ! which kustomize 2>/dev/null ; then
 
-for IMAGE_VAR in IRONIC_IMAGE IRONIC_INSPECTOR_IMAGE ; do
+# Pull images we'll need
+for IMAGE_VAR in IPA_DOWNLOADER_IMAGE VBMC_IMAGE SUSHY_TOOLS_IMAGE; do
```
adding a variable with the images would be nice

elfosardo commented on 01_prepare_host.sh:
```diff
@@ -38,20 +38,36 @@ if ! which kustomize 2>/dev/null ; then
 
-## Download centos 7 qcow2 image
+# Download IPA and CentOS 7 Images
+mkdir -p "$IRONIC_DATA_DIR/html/images"
```
I think the path can be change to a vaiable since it's reused a couple of times ?

stbenjam commented:
/test-centos-integration

stbenjam commented:
/test-centos-integration

stbenjam commented:
/test-centos-integration

maelk commented:
To fix the interface issue, something similar to https://github.com/metal3-io/metal3-dev-env/pull/71 can be done

stbenjam commented on 01_prepare_host.sh:
```diff
@@ -60,15 +76,14 @@ if [ ! -f ${CENTOS_IMAGE} ] ; then
   -b -vvv vm-setup/install-package-playbook.yml
 
-sudo "${CONTAINER_RUNTIME}" pull "${VBMC_IMAGE}"
-sudo "${CONTAINER_RUNTIME}" pull "${SUSHY_TOOLS_IMAGE}"
```
It's just moved into the for loop above where we pre-load the others, line # 57

stbenjam commented:
/test-centos-integration

russellb commented:
/test-centos-integration

hardys commented on Makefile:
```diff
@@ -15,7 +15,7 @@ verify:
 
 delete_mgmt_cluster:
-	sudo minikube delete || true
+	sudo su -l -c "minikube delete" "$(USER)"
```
I noticed we're not deleting the minikube VM too but I fixed it by just removing the sudo?

hardys replied:
Hmm actually I see this matches the sudo in `03_launch_mgmt_cluster.sh` - I wonder if we actually need sudo in either place provided the user is in the libvirt group?

stbenjam replied:
I'm not sure, I just noticed it wasn't being deleted and made it consistent with everywhere else we're running minikube in metal3-dev-env. I assume if the user is a member of libivrt it would work fine, but maybe this is because of support Ubuntu?

maelk replied:
The issue is that if you run the whole script in an ubuntu or centos VM that run an off-the-shelf image, the group is created and the user added, but it does not take effect immediately. So not using this sudo su trick would mean that the scripts would fail until you log out and in again.

stbenjam commented:
/test-integration

stbenjam commented:
This is passing, once https://github.com/metal3-io/baremetal-operator/pull/309 is merged I can point this PR back at master and squash my commits.

russellb commented:
> This is passing, once [metal3-io/baremetal-operator#309](https://github.com/metal3-io/baremetal-operator/pull/309) is merged I can point this PR back at master and squash my commits.

looks like we'll need to merge them at the same time.  I'm going to do another pass of review, and then we can sync up on slack to get them in together.

stbenjam commented:
/test-centos-integration

stbenjam commented:
/test-integration

stbenjam commented:
Last CI run passed: https://jenkins.nordix.org/job/airship_metal3io_metal3_dev_env_integration_test_ubuntu/34/

https://jenkins.nordix.org/job/airship_metal3io_metal3_dev_env_integration_test_centos/27/

I've squashed everything, and moved the code back to hitting metal3-io/baremetal-operator instead of my fork.

maelk commented on 04_verify.sh:
```diff
@@ -227,12 +227,5 @@ if [[ "${BMO_RUN_LOCAL}" == true ]] || [[ "${CAPBM_RUN_LOCAL}" == true ]]; then
 fi
 
-#Verify Ironic containers are running
-for name in ironic ironic-inspector dnsmasq httpd mariadb; do
```
httpd is still running on the host, we could keep the check for this container

stbenjam replied:
Good point, added the check back

maelk commented on 01_prepare_host.sh:
```diff
@@ -38,20 +38,31 @@ if ! command -v kustomize 2>/dev/null ; then
+esac
+
+# Pull images we'll need
+for IMAGE_VAR in IPA_DOWNLOADER_IMAGE VBMC_IMAGE SUSHY_TOOLS_IMAGE; do
```
This is missing ironic image, as it is used for httpd container

stbenjam replied:
Fixed

russellb commented:
merged the baremetal-operator PR, so let's run CI on this one more time ...

/test-centos-integration

russellb commented:
@stbenjam Can you prepare a docs PR against metal3-io/metal3-io.github.io that reflects the changes to how you would talk to the Ironic API?  See the "try it" page contents.

stbenjam commented:
/test-centos-integration

russellb commented:
/hold

Just seeing if prow is listening now ...

russellb commented:
Yay

/hold cancel

stbenjam commented:
/test-integration

stbenjam commented:
/test-integration

maelk commented:
Can we merge this ? Metal3-dev-env is broken without this.

stbenjam commented:
@maelk Sure, I think so, anything else we can address in follow-ups. If you do `/lgtm` tide should merge this automatically

maelk commented:
/lgtm
